Libplanet/Net/Swarm.TrustedStateCompleter.cs

```diff
@@ -0,0 +1,147 @@
+            CancellationToken cancellationToken = default(CancellationToken)
+        )
+        {
+            if (trustedStateValidators is null)
+            {
+                throw new ArgumentNullException(nameof(trustedStateValidators));
+            }
+
+            IReadOnlyList<BoundPeer> trustedPeers =
+                (await DialToExistingPeers(dialTimeout, cancellationToken))
+                    .Select(pair => pair.Peer)
+                    .ToArray();
+
+            StateCompleterSet<T> fallback = fallbackCompleterSet
+                ?? StateCompleterSet<T>.Recalculate;
+
+            async Task FillTrustedBlockStates(HashDigest<SHA256> blockHash)
+            {
+                var request = new GetBlockStates(blockHash);
+                foreach (BoundPeer peer in trustedPeers)
+                {
+                    _logger.Debug(
+                        "Requests the block {BlockHash}'s states to a trusted peer {Peer}...",
+                        blockHash,
+                        peer
+                    );
+                    Message reply;
+                    try
+                    {
+                        reply = await Transport.SendMessageWithReplyAsync(
+                            peer,
+                            request,
+                            timeout: Options.RecentStateRecvTimeout,
+                            cancellationToken: cancellationToken
+                        );
+                    }
+                    catch (TimeoutException e)
+                    {
+                        string msg = "Failed to receive the block {BlockHash}'s states from " +
+                            "a trusted peer {Peer}: " + e;
+                        _logger.Error(e, msg, blockHash, peer);
+                        continue;
+                    }
+
+                    if (reply is BlockStates blockStates)
+                    {
+                        if (!blockStates.BlockHash.Equals(blockHash))
+                        {
+                            string msg =
+                                $"Expected a {nameof(BlockStates)} message for the block " +
+                                "{RequestedBlockHash}, but received a wrong reply instead from " +
+                                "a trusted peer {Peer}: {RepliedBlockHash}";
+                            _logger.Error(msg, blockHash, peer, blockStates.BlockHash);
+                            continue;
+                        }
+                        else if (blockStates.States is null)
+                        {
+                            string msg =
+                                "A trusted peer {Peer} does not have the block {BlockHash}'s " +
+                                "states.";
+                            _logger.Error(msg, peer, blockHash);
+                            continue;
+                        }
+
+                        BlockChain.Store.SetBlockStates(blockStates.BlockHash, blockStates.States);
```

Comment:
Is it fine without any of the mutex?

Reply:
Rather it does not work if we try to lock the `BlockChain._rwlock` here, because `BlockChain<T>.GetState()`/`GetBalance()` methods guarantee their `stateCompleter` to be called only in the read lock:

https://github.com/planetarium/libplanet/blob/ed9b1f86532d3440176511b68da55b795ed6fca9/Libplanet/Blockchain/BlockChain.cs#L1680-L1702

Reply:
I see... then it seems that we should upgrade `_rwlock` before calling `rawStateCompleter` in `BlockChain<T>.GetRawState()`.

Reply:
Not all state completers update the blockchain, but only `Recalculate` completers do.  Those implementations have already entered the write lock by themselves.

<https://github.com/planetarium/libplanet/blob/ed9b1f86532d3440176511b68da55b795ed6fca9/Libplanet/Blockchain/BlockChain.cs#L1638-L1647>

Reply:
Then, shouldn't these state completers upgrade `_rwlocks`? (like Recalculate)

Reply:
@longfin I'm going to address this problem in the next separated pull request!